npm/private/utils.bzl

```diff
@@ -14,70 +14,97 @@ def _sorted_map(m):
+                r += "+"
+        elif c == "@" and i > version_index:
+            if r and r[-1] != "_":
+                # use "_" for version @ in peers
+                r += "_"
+            if s[i - 1] == "(":
+                # use "at" for scope @ in peers
+                r += "at_"
+        elif not c.isalnum() and c != "-" and c != "_" and c != "." and c != "+" and c != "@":
+            r += "_"
+        else:
+            r += c
+
+    r = r.rstrip("_-.")
+
+    return r
+
+def _package_repo_name(prefix, s):
     # Workspace names may contain only A-Z, a-z, 0-9, '-', '_' and '.'
-    result = ""
-    for c in string.elems():
-        if c == "@" and (not result or result[-1] == "_"):
-            result += "at"
-        if not c.isalnum() and c != "-" and c != "_" and c != ".":
-            c = "_"
-        result += c
-    return result.strip("_-")
-
-def _bazel_name(name, version = None):
-    "Make a bazel friendly name from a package name and (optionally) a version that can be used in repository and target names"
-    escaped_name = _sanitize_rule_name(name)
-    if not version:
-        return escaped_name
-
-    # Separate name + version with extra _
-    return "%s__%s" % (escaped_name, _sanitize_rule_name(version))
-
-def _package_key(name, version):
-    "Make a name/version pnpm-style name for a package name and version"
-    return "%s@%s" % (name, version)
+    # Package repo names handle '/' and other characters unique to better represent npm packages.
 
-def _friendly_name(name, version):
-    "Make a name@version developer-friendly name for a package name and version"
-    return "%s@%s" % (name, version)
+    # TODO(3.0): simplify and stop trying to be backwards compatible with 2.x
 
-def _escape_target_name(name):
-    return name.replace("://", "/").replace("/", "+").replace(":", "+")
+    version_index = s.find("@", 1)
+    peers_index = s.find("(", version_index)
 
-def _package_store_name(pnpm_name, pnpm_version):
-    "Make a package store name for a given package and version"
+    peer_suffix = ""
+    if peers_index != -1:
+        peer_dep = s[peers_index:]
+        s = s[:peers_index]
 
-    if pnpm_version.startswith("link:"):
-        # Distinguish local links a 0.0.0 version. This is unlike pnpm which symlinks
-        # local links into the source tree instead of storing them in the package store.
-        name = pnpm_name
-        version = "0.0.0"
-    elif pnpm_version.startswith("npm:"):
-        name, version = pnpm_version[4:].rsplit("@", 1)
-    else:
-        name = pnpm_name
-        version = pnpm_version
-
-    # TODO(3.0): the version should already be the store name to remove this pnpm lockfile
-    # logic of knowing/guessing when a version is already a store name.
-
-    at_index = version.find("@")
-    proto_indx = version.find("://")
-    if at_index == 0:
-        # Special case where the package name should _not_ be included in the package store name.
```

Comment:
This was also one of the major issues... at this point we just had name+version and we don't truly know what the proper store name is. We were just guessing based on assumptions originally from pnpm 5.